text/0131-pass-native-sdk-spans-to-hybrid.md

```diff
@@ -0,0 +1,174 @@
+    When the native SpansAggregator has to flush all its spans
+    Then native SDK pushes all native spans up to the hybrid SDK
+    And doesn't send its native spans directly to Sentry
+    And the hybrid SDK adds all native spans to its SpansAggregator
+
+Scenario: Active span, native SpansAggregator flushes
+    Given an active span
+    When the native SpansAggregator has to flush all its spans
+    Then native SDK pushes all native spans up to the hybrid SDK
+    And the hybrid SDK filters for spans belonging to the active span 
+    And adds these spans to the active span
+    And adds the non matching spans to the hybrid SpansAggregator
+
+Scenario: Active span finishes
+    Given an active span
+    When it finishes
+    Then hybrid SDK gets all spans from the native SpansAggregator
+    And the hybrid SDK filters for spans belonging to the active span 
+    And adds these spans to the active span
+    And adds the non matching spans to the hybrid SpansAggregator
+
+Scenario: Active span, hybrid SpansAggregator flushes
+    Given an active span
+    When the hybrid SpansAggregator has to flush all its spans
+    Then it does not pull spans from the native SpansAggregator
+    And sends all hybrid spans from the SpansAggregator to Sentry
+
+Scenario: No active span, hybrid SpansAggregator flushes
+    Given no active span
+    When the hybrid SpansAggregator has to flush all its spans
+    Then it pulls spans from the native SpansAggregator
+    And adds these to the hybrid SpansAggregator
+    And sends native and hybrid spans to Sentry
+```
+
+### Pros <a name="option-2-pros"></a>
+
+1. Reusing existing functionality of the SpansAggregator.
+2. Works well with single-span ingestion and transactions.
+3. Adding spans to the SpansAggregator is `O(1)`.
+4. Less cross hybrid layer communication than [option 3](#option-3).
+
+### Cons <a name="option-2-cons"></a>
+
+1. Please add your cons.
+
+
+## Option 3: Native Callbacks<a name="option-3"></a>
+
+The native layers offer callbacks whenever they start and finish auto-instrumented spans and don't
+keep track of spans. For example, the HTTP request instrumentation offers two callbacks: one for
+starting the HTTP request and another for finishing the HTTP request. The hybrid SDKs then subscribe
+to these callbacks and create the spans themselves.
+
+### Pros <a name="option-3-pros"></a>
+
+1. This option works well with idle transactions and `waitForChildren` because these concepts need to
+know when spans start and finish, but this is a weak argument as with single-span ingestion, this
+concept will become obsolete.
+2. Works well with single-span ingestion, as the hybrid SDKs get callbacks for each span.
+3. Complete flexibility for hybrid SDKs. Hybrid SDKs have full control over how they want to deal with
+native-span information.
+
+### Cons <a name="option-3-cons"></a>
+
+1. Each span requires two calls across the layers, which could negatively impact performance.
```

Comment:
What prevent us from having only two callbacks and having arguments to control what kind o span was started/finished and its properties?

Reply:
I'm sorry, I don't fully understand. Do you maybe mean one callback? 

> arguments to control what kind o span was started/finished and its properties

An interesting idea 🤔. How would that work? What are the benefits of that?

Reply:
You mentioned "each spans requires two" I guess http spans requires two, db spans requires two, file I/O requires two. 

I suggested two callbacks for all of this, along with "hints" to pass information such as "this is an HTTP request to URL X". However, we could even just use one callback and another argument to indicate the start or finish.

Reply:
Ah, yes, obviously, this makes sense. Two or one callback makes sense.